modulemd/v2/modulemd-module-index.c

```diff
@@ -0,0 +1,536 @@
+          return FALSE;
+        }
+    }
+
+  return TRUE;
+}
+
+
+static gboolean
+modulemd_module_index_dump_to_emitter (ModulemdModuleIndex *self,
+                                       yaml_emitter_t *emitter,
+                                       GError **error)
+{
+  GHashTableIter iter;
+  ModulemdModule *module = NULL;
+  gpointer value;
+
+  if (!mmd_emitter_start_stream (emitter, error))
+    return FALSE;
+
+  g_hash_table_iter_init (&iter, self->modules);
+
+  while (g_hash_table_iter_next (&iter, NULL, &value))
+    {
+      module = (ModulemdModule *)value;
+
+      if (!dump_defaults (module, emitter, error))
+        return FALSE;
+
+      if (!dump_translations (module, emitter, error))
+        return FALSE;
+
+      if (!dump_streams (module, emitter, error))
+        return FALSE;
+    }
+
+  if (!mmd_emitter_end_stream (emitter, error))
+    return FALSE;
+
+  return TRUE;
+}
+
+
+gboolean
+modulemd_module_index_update_from_file (ModulemdModuleIndex *self,
+                                        const gchar *yaml_file,
+                                        GPtrArray **failures,
+                                        GError **error)
+{
+  g_return_val_if_fail (MODULEMD_IS_MODULE_INDEX (self), FALSE);
+
+  g_autoptr (FILE) yaml_stream = NULL;
+
+  yaml_stream = g_fopen (yaml_file, "rb");
+
+  return modulemd_module_index_update_from_stream (
+    self, yaml_stream, failures, error);
+}
+
+
+gboolean
+modulemd_module_index_update_from_string (ModulemdModuleIndex *self,
+                                          const gchar *yaml_string,
+                                          GPtrArray **failures,
+                                          GError **error)
+{
+  g_return_val_if_fail (MODULEMD_IS_MODULE_INDEX (self), FALSE);
+
+  MMD_INIT_YAML_PARSER (parser);
+
+  yaml_parser_set_input_string (
+    &parser, (const unsigned char *)yaml_string, strlen (yaml_string));
+
+  return modulemd_module_index_update_from_parser (
+    self, &parser, failures, error);
+}
+
+
+gboolean
+modulemd_module_index_update_from_stream (ModulemdModuleIndex *self,
+                                          FILE *yaml_stream,
+                                          GPtrArray **failures,
+                                          GError **error)
+{
+  g_return_val_if_fail (MODULEMD_IS_MODULE_INDEX (self), FALSE);
+
+  MMD_INIT_YAML_PARSER (parser);
+
+  yaml_parser_set_input_file (&parser, yaml_stream);
+
+  return modulemd_module_index_update_from_parser (
+    self, &parser, failures, error);
+}
+
+
+const gchar *
+modulemd_module_index_dump_to_string (ModulemdModuleIndex *self,
+                                      GError **error)
+{
+  g_return_val_if_fail (MODULEMD_IS_MODULE_INDEX (self), NULL);
+
+  MMD_INIT_YAML_EMITTER (emitter);
+  MMD_INIT_YAML_STRING (&emitter, yaml_string);
+
+  if (!modulemd_module_index_dump_to_emitter (self, &emitter, error))
+    return NULL;
+
+  return g_strdup (yaml_string->str);
```

Comment:
Use `g_steal_pointer()` here instead. It's likely to be a very large string and there's no need to copy it as it's about to be freed anyway.